gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/pooled_grpc_asyncio.py.j2

```diff
@@ -0,0 +1,445 @@
+        {% endif %}
+
+        if api_mtls_endpoint:
+            warnings.warn("api_mtls_endpoint is deprecated", DeprecationWarning)
+        if client_cert_source:
+            warnings.warn("client_cert_source is deprecated", DeprecationWarning)
+
+        if api_mtls_endpoint:
+            host = api_mtls_endpoint
+
+            # Create SSL credentials with client_cert_source or application
+            # default SSL credentials.
+            if client_cert_source:
+                cert, key = client_cert_source()
+                self._ssl_channel_credentials = grpc.ssl_channel_credentials(
+                    certificate_chain=cert, private_key=key
+                )
+            else:
+                self._ssl_channel_credentials = SslCredentials().ssl_credentials
+
+        else:
+            if client_cert_source_for_mtls and not ssl_channel_credentials:
+                cert, key = client_cert_source_for_mtls()
+                self._ssl_channel_credentials = grpc.ssl_channel_credentials(
+                    certificate_chain=cert, private_key=key
+                )
+
+        # The base transport sets the host, credentials and scopes
+        super().__init__(
+            host=host,
+            credentials=credentials,
+            credentials_file=credentials_file,
+            scopes=scopes,
+            quota_project_id=quota_project_id,
+            client_info=client_info,
+            always_use_jwt_access=always_use_jwt_access,
+            api_audience=api_audience,
+        )
+        self._quota_project_id = quota_project_id
+        self.channel_pool: List[aio.Channel] = []
+        for i in range(pool_size):
+            new_channel = type(self).create_channel(
+                self._host,
+                # use the credentials which are saved
+                credentials=self._credentials,
+                # Set ``credentials_file`` to ``None`` here as
+                # the credentials that we saved earlier should be used.
+                credentials_file=None,
+                scopes=self._scopes,
+                ssl_credentials=self._ssl_channel_credentials,
+                quota_project_id=self._quota_project_id,
+                options=[
+                    ("grpc.max_send_message_length", -1),
+                    ("grpc.max_receive_message_length", -1),
+                ],
+            )
+            self.channel_pool.append(new_channel)
+
+        # Wrap messages. This must be done after self.channel_pool is populated
+        self._prep_wrapped_messages(client_info)
+
+    def next_channel(self) -> aio.Channel:
+        """Returns the next channel in the round robin pool."""
+        # Return the channel from cache.
+        channel = self.channel_pool[self._next_idx]
+        self._next_idx = (self._next_idx + 1) % len(self.channel_pool)
+        return channel
+
+    async def replace_channel(self, channel_idx, grace=None, new_channel=None) -> aio.Channel:
```

Comment:
Is this method to be called by the client internally, or by the user?

Reply:
From the transport side, this will be user-exposed, but it will probably be client code calling this, not end users.

For BigTable, we need to call this to refresh channels before they expire, which causes a latency spike. We also need to pass in custom channels which have called cache warm-up RPCs, and will also likely use this to add emulator channels. 

I think if other libraries end up using pooling, they may also have other reasons for swapping channels in the pool